README
The Python Imaging Library
$Id$

Release 1.1.7+ (November 15, 2009)

====================================================================
The Python Imaging Library 1.1.7+
====================================================================

Contents
--------

+ Introduction
+ Support Options
  - Commercial support
  - Free support
+ Software License
+ Build instructions (all platforms)
  - Additional notes for Mac OS X
  - Additional notes for Windows

--------------------------------------------------------------------
Introduction
--------------------------------------------------------------------

The Python Imaging Library (PIL) adds image processing capabilities
to your Python environment.  This library provides extensive file
format support, an efficient internal representation, and powerful
image processing capabilities.

This source kit has been built and tested with Python 2.0 and newer,
on Windows, Mac OS X, and major Unix platforms.  Large parts of the
library also work on 1.5.2 and 1.6.

The main distribution site for this software is:

        http://www.pythonware.com/products/pil/

That site also contains information about free and commercial support
options, PIL add-ons, answers to frequently asked questions, and more.


Development versions (alphas, betas) are available here:

        http://effbot.org/downloads/


The PIL handbook is not included in this distribution; to get the
latest version, check:

        http://www.pythonware.com/library/
        http://effbot.org/books/imagingbook/ (drafts)


For installation and licensing details, see below.


--------------------------------------------------------------------
Support Options
--------------------------------------------------------------------

+ Commercial Support

Secret Labs (PythonWare) offers support contracts for companies using
the Python Imaging Library in commercial applications, and in mission-
critical environments.  The support contract includes technical support,
bug fixes, extensions to the PIL library, sample applications, and more.

For the full story, check:

        http://www.pythonware.com/products/pil/support.htm


+ Free Support

For support and general questions on the Python Imaging Library, send
e-mail to the Image SIG mailing list:

        image-sig@python.org

You can join the Image SIG by sending a mail to:

        image-sig-request@python.org

Put "subscribe" in the message body to automatically subscribe to the
list, or "help" to get additional information.  Alternatively, you can
send your questions to the Python mailing list, python-list@python.org,
or post them to the newsgroup comp.lang.python.  DO NOT SEND SUPPORT
QUESTIONS TO PYTHONWARE ADDRESSES.


--------------------------------------------------------------------
Software License
--------------------------------------------------------------------

The Python Imaging Library is

Copyright (c) 1997-2009 by Secret Labs AB
Copyright (c) 1995-2009 by Fredrik Lundh

By obtaining, using, and/or copying this software and/or its
associated documentation, you agree that you have read, understood,
and will comply with the following terms and conditions:

Permission to use, copy, modify, and distribute this software and its
associated documentation for any purpose and without fee is hereby
granted, provided that the above copyright notice appears in all
copies, and that both that copyright notice and this permission notice
appear in supporting documentation, and that the name of Secret Labs
AB or the author not be used in advertising or publicity pertaining to
distribution of the software without specific, written prior
permission.

SECRET LABS AB AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH REGARD TO
THIS SOFTWARE, INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND
FITNESS.  IN NO EVENT SHALL SECRET LABS AB OR THE AUTHOR BE LIABLE FOR
ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT
OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.


--------------------------------------------------------------------
Build instructions (all platforms)
--------------------------------------------------------------------

For a list of changes in this release, see the CHANGES document.

0. If you're in a hurry, try this:

        $ tar xvfz Imaging-1.1.7.tar.gz
        $ cd Imaging-1.1.7
        $ python setup.py install

   If you prefer to know what you're doing, read on.


1. Prerequisites.

   If you need any of the features described below, make sure you
   have the necessary libraries before building PIL.

   feature              library
   -----------------------------------------------------------------
   JPEG support         libjpeg (6b or 8b)

                        http://www.ijg.org
                        http://www.ijg.org/files/jpegsrc.v6b.tar.gz
                        http://www.ijg.org/files/jpegsrc.v8b.tar.gz
                        ftp://ftp.uu.net/graphics/jpeg/

   PNG support          zlib (1.2.3 or later is recommended)

                        http://www.gzip.org/zlib/

   OpenType/TrueType    freetype2 (2.3.9 or later is recommended)
   support              
                        http://www.freetype.org
                        http://freetype.sourceforge.net  

   CMS support          littleCMS (1.1.5 or later is recommended)
   support              
                        http://www.littlecms.com/

   If you have a recent Linux version, the libraries provided with the
   operating system usually work just fine.  If some library is
   missing, installing a prebuilt version (jpeg-devel, zlib-devel,
   etc) is usually easier than building from source.  For example, for
   Ubuntu 9.10 (karmic), you can install the following libraries:

       sudo apt-get install libjpeg62-dev
       sudo apt-get install zlib1g-dev
       sudo apt-get install libfreetype6-dev
       sudo apt-get install liblcms1-dev

   If you're using Mac OS X, you can use the 'fink' tool to install
   missing libraries (also see the Mac OS X section below).

   Similar tools are available for many other platforms.


2. To build under Python 1.5.2, you need to install the stand-alone
   version of the distutils library:

       http://www.python.org/sigs/distutils-sig/download.html

   You can fetch distutils 1.0.2 from the Python source repository:
   
       svn export http://svn.python.org/projects/python/tags/Distutils-1_0_2/Lib/distutils/

   For newer releases, the distutils library is included in the
   Python standard library.

   NOTE: Version 1.1.7 is not fully compatible with 1.5.2.  Some
   more recent additions to the library may not work, but the core
   functionality is available.
   

3. If you didn't build Python from sources, make sure you have
   Python's build support files on your machine.  If you've down-
   loaded a prebuilt package (e.g. a Linux RPM), you probably
   need additional developer packages.  Look for packages named
   "python-dev", "python-devel", or similar.  For example, for
   Ubuntu 9.10 (karmic), use the following command:

       sudo apt-get install python-dev


4. When you have everything you need, unpack the PIL distribution
   (the file Imaging-1.1.7.tar.gz) in a suitable work directory:

        $ cd MyExtensions # example
        $ gunzip Imaging-1.1.7.tar.gz
        $ tar xvf Imaging-1.1.7.tar


5. Build the library.  We recommend that you do an in-place build,
   and run the self test before installing.

        $ cd Imaging-1.1.7
        $ python setup.py build_ext -i
        $ python selftest.py

   During the build process, the setup.py will display a summary
   report that lists what external components it found.  The self-
   test will display a similar report, with what external components
   the tests found in the actual build files:

        ----------------------------------------------------------------
        PIL 1.1.7 SETUP SUMMARY
        ----------------------------------------------------------------
        *** TKINTER support not available (Tcl/Tk 8.5 libraries needed)
        --- JPEG support available
        --- ZLIB (PNG/ZIP) support available
        --- FREETYPE support available
        ----------------------------------------------------------------

   Make sure that the optional components you need are included.

   If the build script won't find a given component, you can edit the
   setup.py file and set the appropriate ROOT variable.  For details,
   see instructions in the file.

   If the build script finds the component, but the tests cannot
   identify it, try rebuilding *all* modules:

        $ python setup.py clean
        $ python setup.py build_ext -i


6. If the setup.py and selftest.py commands finish without any
   errors, you're ready to install the library:

        $ python setup.py install

   (depending on how Python has been installed on your machine,
   you might have to log in as a superuser to run the 'install'
   command, or use the 'sudo' command to run 'install'.)


--------------------------------------------------------------------
Additional notes for Mac OS X
--------------------------------------------------------------------

On Mac OS X you will usually install additional software such as
libjpeg or freetype with the "fink" tool, and then it ends up in
"/sw".  If you have installed the libraries elsewhere, you may have
to tweak the "setup.py" file before building.


--------------------------------------------------------------------
Additional notes for Windows
--------------------------------------------------------------------

On Windows, you need to tweak the ROOT settings in the "setup.py"
file, to make it find the external libraries.  See comments in the
file for details.

Make sure to build PIL and the external libraries with the same
runtime linking options as was used for the Python interpreter
(usually /MD, under Visual Studio).


Note that most Python distributions for Windows include libraries
compiled for Microsoft Visual Studio.  You can get the free Express
edition of Visual Studio from:

    http://www.microsoft.com/Express/

To build extensions using other tool chains, see the "Using
non-Microsoft compilers on Windows" section in the distutils handbook:

    http://www.python.org/doc/current/inst/non-ms-compilers.html

For additional information on how to build extensions using the
popular MinGW compiler, see:

    http://mingw.org (compiler)
    http://sebsauvage.net/python/mingw.html (build instructions)
    http://sourceforge.net/projects/gnuwin32 (prebuilt libraries)



CHANGES
-*- coding: utf-8 -*-

The Python Imaging Library
$Id$

ACKNOWLEDGEMENTS: PIL wouldn't be what it is without the help of:
David Ascher, Phil Austin, Douglas Bagnall, Larry Bates, Anthony
Baxter, William Baxter, Denis Benoit, Jan Blom, Duncan Booth, Alexey
Borzenkov, Jeff Breidenbach, Roger Burnham, Zac Burns, Gene Cash,
Kevin Cazabon, Fred Clare, Greg Coats, Chris Cogdon, Greg Couch, Bill
Crutchfield, Abel Deuring, Tim Docker, Fred Drake, Graham Dumpleton,
Matthew Ellis, Eric Etheridge, Daniel Fetchinson, Robin Friedrich,
Pier Paolo Glave, Federico Di Gregorio, Markus Gritsch, Daniel
Haertle, Greg Hamilton, Mark Hammond, Bernhard Herzog, Rob Hooft, Bob
Ippolito, Jack Jansen, Bill Janssen, Edward Jones, Richard Jones,
Håkan Karlsson, Robert Kern, David Kirtley, Bob Klimek, Matthias
Klose, Andrew Kuchling, Magnus Källström, Victor Lacina, Ben Last,
Hamish Lawson, Cesare Leonardi, Andrew MacIntyre, Jan Matejek, Naveen
Michaud-Agrawal, Gordon McMillan, Skip Montanaro, Fredrik Nehr,
Russell Nelson, Luciano Nocera, Travis Oliphant, Piet van Oostrum,
Richard Oudkerk, Paul Pharr, Andres Polit, Conrado Porto Lopes Gouvêa,
Eric Raymond, Victor Reijs, Bertil Reinhammar, Nicholas Riley, Don
Rozenberg, Toby Sargeant, Barry Scott, Les Schaffer, Joel Shprentz,
Klamer Shutte, Gene Skonicki, Niki Spahiev, D. Alan Stewart, Perry
Stoll, Paul Svensson, Ulrik Svensson, Miki Tebeka, Michael van
Tellingen, Ivan Tkatchev, Dan Torop, Adam Twardoch, Rune Uhlin, Dmitry
Vasiliev, Sasha Voynow, Charles Waldman, Collin Winter, Dan Wolfe,
Ka-Ping Yee, and many others (if your name should be on this list, let
me know.)

*** Changes from release 1.1.7 to 1.1.8 ***

+ Added support for reading 16-bit RGB TIFF file.

*** Changes from release 1.1.6 to 1.1.7 ***

This section may not be fully complete.  For changes since this file
was last updated, see the repository revision history:

  http://bitbucket.org/effbot/pil-2009-raclette/changesets/

(1.1.7 final)

+ Set GIF loop info property to the number of iterations if a NETSCAPE
  loop extension is present, instead of always setting it to 1 (from
  Valentino Volonghi).

(1.1.7c1 released)

+ Improved PNG compression (from Alexey Borzenkov).

+ Read interlaced PNG files (from Conrado Porto Lopes Gouvêa)

+ Added various TGA improvements from Alexey Borzenkov, including
  support for specifying image orientation.

+ Bumped block threshold to 16 megabytes, made size estimation a bit
  more accurate.  This speeds up allocation of large images.

+ Fixed rounding error in ImagingDrawWideLine.

  "gormish" writes: ImagingDrawWideLine() in Draw.c has a bug in every
  version I've seen, which leads to different width lines depending on
  the order of the points in the line. This is especially bad at some
  angles where a 'width=2' line can completely disappear.

+ Added support for RGBA mode to the SGI module (based on code by
  Karsten Hiddemann).

+ Handle repeated IPTC tags (adapted from a patch by Eric Bruning).

  Eric writes: According to the specification, some IPTC tags can be
  repeated, e.g., tag 2:25 (keywords). PIL 1.1.6 only retained the last
  instance of that tag. Below is a patch to store all tags. If there are
  multiple tag instances, they are stored in a (python) list. Single tag
  instances remain as strings.

+ Fixed potential crash in ImageFilter for small target images
  (reported by Zac Burns and Daniel Fetchinson).

+ Use BMP instead of JPEG as temporary show format on Mac OS X.

+ Fixed putpixel/new for I;16 with colors > 255.

+ Added integer power support to ImagingMath.

+ Added limited support for I;16L mode (explicit little endian).

+ Moved WMF support into Image.core; enable WMF rendering by default
  if renderer is available.

+ Mark the ARG plugin as obsolete.

+ Added version query mechanism to ImageCms and ImageFont, for
  debugging.

+ Added (experimental) ImageCms function for fetching the ICC profile
  for the current display (currently Windows only).

  Added HWND/HDC support to ImageCms.get_display_profile().

+ Added WMF renderer (Windows only).

+ Added ImagePointHandler and ImageTransformHandler mixins; made
  ImageCmsTransform work with im.point.

+ Fixed potential endless loop in the XVThumbnail reader (from Nikolai
  Ugelvik).

+ Added Kevin Cazabon's pyCMS package.

  The C code has been moved to _imagingcms.c, the Python interface
  module is installed as PIL.ImageCMS.

  Added support for in-memory ICC profiles.

  Unified buildTransform and buildTransformFromOpenProfiles.

  The profile can now be either a filename, a profile object, or a
  file-like object containing an in-memory profile.

  Additional fixes from Florian Böch:

    Very nice - it just needs LCMS flags support so we can use black
    point compensation and softproofing :) See attached patches.  They
    also fix a naming issue which could cause confusion - display
    profile (ImageCms wording) actually means proof profile (lcms
    wording), so I changed variable names and docstrings where
    applicable. Patches are tested under Python 2.6.

+ Improved support for layer names in PSD files (from Sylvain Baubeau)

  Sylvain writes: I needed to be able to retrieve the names of the
  layers in a PSD files. But PsdImagePlugin.py didn't do the job so I
  wrote this very small patch.

+ Improved RGBA support for ImageTk for 8.4 and newer (from Con
  Radchenko).

  This replaces the slow run-length based encoding model with true
  compositing at the Tk level.

+ Added support for 16- and 32-bit images to McIdas loader.

  Based on file samples and stand-alone reader code provided by Craig
  Swank.

+ Added ImagePalette support to putpalette.

+ Fixed problem with incremental parsing of PNG files.

+ Make selftest.py report non-zero status on failure (from Mark
  Sienkiewicz)

+ Add big endian save support and multipage infrastructure to the TIFF
  writer (from Sebastian Haase).

+ Handle files with GPS IFD but no basic EXIF IFD (reported by Kurt
  Schwehr).


selftest.py
# minimal sanity check

ROOT = "."

import os, sys
sys.path.insert(0, ROOT)

from PIL import Image
from PIL import ImageDraw
from PIL import ImageFilter
from PIL import ImageMath

try:
    Image.core.ping
except ImportError as v:
    print("***", v)
    sys.exit()
except AttributeError:
    pass

def _info(im):
    im.load()
    return im.format, im.mode, im.size

def testimage():
    """
    PIL lets you create in-memory images with various pixel types:

    >>> im = Image.new("1", (128, 128)) # monochrome
    >>> _info(im)
    (None, '1', (128, 128))
    >>> _info(Image.new("L", (128, 128))) # grayscale (luminance)
    (None, 'L', (128, 128))
    >>> _info(Image.new("P", (128, 128))) # palette
    (None, 'P', (128, 128))
    >>> _info(Image.new("RGB", (128, 128))) # truecolor
    (None, 'RGB', (128, 128))
    >>> _info(Image.new("I", (128, 128))) # 32-bit integer
    (None, 'I', (128, 128))
    >>> _info(Image.new("F", (128, 128))) # 32-bit floating point
    (None, 'F', (128, 128))

    Or open existing files:

    >>> im = Image.open(os.path.join(ROOT, "Images/lena.gif"))
    >>> _info(im)
    ('GIF', 'P', (128, 128))
    >>> _info(Image.open(os.path.join(ROOT, "Images/lena.ppm")))
    ('PPM', 'RGB', (128, 128))
    >>> try:
    ...  _info(Image.open(os.path.join(ROOT, "Images/lena.jpg")))
    ... except IOError as v:
    ...  print(v)
    ('JPEG', 'RGB', (128, 128))

    PIL doesn't actually load the image data until it's needed,
    or you call the "load" method:

    >>> im = Image.open(os.path.join(ROOT, "Images/lena.ppm"))
    >>> print(im.im) # internal image attribute
    None
    >>> a = im.load()
    >>> str(type(im.im)).split("'")[1]
    'ImagingCore'

    You can apply many different operations on images.  Most
    operations return a new image:

    >>> im = Image.open(os.path.join(ROOT, "Images/lena.ppm"))
    >>> _info(im.convert("L"))
    (None, 'L', (128, 128))
    >>> _info(im.copy())
    (None, 'RGB', (128, 128))
    >>> _info(im.crop((32, 32, 96, 96)))
    (None, 'RGB', (64, 64))
    >>> _info(im.filter(ImageFilter.BLUR))
    (None, 'RGB', (128, 128))
    >>> im.getbands()
    ('R', 'G', 'B')
    >>> im.getbbox()
    (0, 0, 128, 128)
    >>> len(im.getdata())
    16384
    >>> im.getextrema()
    ((61, 255), (26, 234), (44, 223))
    >>> im.getpixel((0, 0))
    (223, 162, 133)
    >>> len(im.getprojection())
    2
    >>> len(im.histogram())
    768
    >>> _info(im.point(list(range(256))*3))
    (None, 'RGB', (128, 128))
    >>> _info(im.resize((64, 64)))
    (None, 'RGB', (64, 64))
    >>> _info(im.rotate(45))
    (None, 'RGB', (128, 128))
    >>> [_info(o) for o in im.split()]
    [(None, 'L', (128, 128)), (None, 'L', (128, 128)), (None, 'L', (128, 128))]
    >>> len(im.convert("1").tobitmap())
    10456
    >>> len(im.tostring())
    49152
    >>> _info(im.transform((512, 512), Image.AFFINE, (1,0,0,0,1,0)))
    (None, 'RGB', (512, 512))
    >>> _info(im.transform((512, 512), Image.EXTENT, (32,32,96,96)))
    (None, 'RGB', (512, 512))

    The ImageDraw module lets you draw stuff in raster images:

    >>> im = Image.new("L", (128, 128), 64)
    >>> d = ImageDraw.ImageDraw(im)
    >>> d.line((0, 0, 128, 128), fill=128)
    >>> d.line((0, 128, 128, 0), fill=128)
    >>> im.getextrema()
    (64, 128)

    In 1.1.4, you can specify colors in a number of ways:

    >>> xy = 0, 0, 128, 128
    >>> im = Image.new("RGB", (128, 128), 0)
    >>> d = ImageDraw.ImageDraw(im)
    >>> d.rectangle(xy, "#f00")
    >>> im.getpixel((0, 0))
    (255, 0, 0)
    >>> d.rectangle(xy, "#ff0000")
    >>> im.getpixel((0, 0))
    (255, 0, 0)
    >>> d.rectangle(xy, "rgb(255,0,0)")
    >>> im.getpixel((0, 0))
    (255, 0, 0)
    >>> d.rectangle(xy, "rgb(100%,0%,0%)")
    >>> im.getpixel((0, 0))
    (255, 0, 0)
    >>> d.rectangle(xy, "hsl(0, 100%, 50%)")
    >>> im.getpixel((0, 0))
    (255, 0, 0)
    >>> d.rectangle(xy, "red")
    >>> im.getpixel((0, 0))
    (255, 0, 0)

    In 1.1.6, you can use the ImageMath module to do image
    calculations.

    >>> im = ImageMath.eval("float(im + 20)", im=im.convert("L"))
    >>> im.mode, im.size
    ('F', (128, 128))

    PIL can do many other things, but I'll leave that for another
    day.  If you're curious, check the handbook, available from:

        http://www.pythonware.com

    Cheers /F
    """


def check_module(feature, module):
    try:
        __import__("PIL." + module)
    except ImportError:
        print("***", feature, "support not installed")
    else:
        print("---", feature, "support ok")

def check_codec(feature, codec):
    if codec + "_encoder" not in dir(Image.core):
        print("***", feature, "support not installed")
    else:
        print("---", feature, "support ok")


if __name__ == "__main__":
    # check build sanity

    exit_status = 0

    print("-"*68)
    print("PIL", Image.VERSION, "TEST SUMMARY ")
    print("-"*68)
    print("Python modules loaded from", os.path.dirname(Image.__file__))
    print("Binary modules loaded from", os.path.dirname(Image.core.__file__))
    print("-"*68)
    check_module("PIL CORE", "_imaging")
    check_module("TKINTER", "_imagingtk")
    check_codec("JPEG", "jpeg")
    check_codec("ZLIB (PNG/ZIP)", "zip")
    check_module("FREETYPE2", "_imagingft")
    check_module("LITTLECMS", "_imagingcms")
    print("-"*68)

    # use doctest to make sure the test program behaves as documented!
    import doctest, selftest
    print("Running selftest:")
    status = doctest.testmod(selftest)
    if status[0]:
        print("*** %s tests of %d failed." % status)
        exit_status = 1
    else:
        print("--- %s tests passed." % status[1])

    sys.exit(exit_status)
